src/big/big_decimal.cr

```diff
@@ -0,0 +1,328 @@
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    numerator, denominator = @value, other.@value
+
+    quotient, remainder = numerator.divmod(denominator)
+    if remainder == ZERO
+      return BigDecimal.new(normalize_quotient(other, quotient), scale)
+    end
+
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < max_div_iterations
+      inner_quotient, inner_remainder = remainder.divmod(denominator)
+      quotient = quotient * TEN + inner_quotient
+      remainder = inner_remainder * TEN
+      i += 1
+    end
+
+    BigDecimal.new(normalize_quotient(other, quotient), scale + i)
+  end
+
+  def <=>(other : BigDecimal) : Int32
+    if @scale > other.scale
+      @value <=> other.scale_to(self).value
+    elsif @scale < other.scale
+      scale_to(other).value <=> other.value
+    else
+      @value <=> other.value
+    end
+  end
+
+  def <=>(other : Int)
+    self <=> BigDecimal.new(other)
+  end
+
+  def ==(other : BigDecimal) : Bool
+    if @scale > other.scale
+      scaled = other.value * power_ten_to(@scale - other.scale)
+      @value == scaled
+    elsif @scale < other.scale
+      scaled = @value * power_ten_to(other.scale - @scale)
+      scaled == other.value
+    else
+      @value == other.value
+    end
+  end
+
+  # Scales a `BigDecimal` to another `BigDecimal`, so they can be
+  # computed easier.
+  def scale_to(new_scale : BigDecimal) : BigDecimal
+    new_scale = new_scale.scale
+
+    if @value == 0
+      BigDecimal.new(0.to_big_i, new_scale)
+    elsif @scale > new_scale
+      scale_diff = @scale - new_scale.to_big_i
+      BigDecimal.new(@value / power_ten_to(scale_diff), new_scale)
+    elsif @scale < new_scale
+      scale_diff = new_scale - @scale.to_big_i
+      BigDecimal.new(@value * power_ten_to(scale_diff), new_scale)
+    else
+      self
+    end
+  end
+
+  def to_s(io : IO)
+    factor_powers_of_ten
+    s = @value.to_s
+
+    if @scale == 0
+      io << s
+      return
+    end
+
+    if @scale >= s.size && @value >= 0
+      io << "0."
+      (@scale - s.size).times do
+        io << '0'
+      end
+      io << s
+    elsif @scale >= s.size && @value < 0
+      io << "-0.0"
+      (@scale - s.size).times do
+        io << '0'
+      end
+      io << s[1..-1]
+    else
+      offset = s.size - @scale
+      io << s[0...offset] << '.' << s[offset..-1]
+    end
+  end
+
+  def to_big_d
+    self
+  end
+
+  # Converts to integer. Truncates anything on the right side of the decimal point.
+  def to_i
+    if @value >= 0
+      (@value / TEN ** @scale).to_i
+    else
+      -(@value.abs / TEN ** @scale).to_i
+    end
+  end
+
+  # Converts to unsigned integer. Truncates anything on the right side of the decimal point,
+  # converting negative to positive.
+  def to_u
+    (@value.abs / TEN ** @scale).to_u
+  end
+
+  def to_f
+    to_s.to_f
+  end
+
+  def clone
+    self
+  end
+
+  def hash(hasher)
+    hasher.string(self.to_s)
```

Comment:
Of course it must be rewritten using number normalization.

But it will be anyway later.